core/src/main/scala/org/apache/spark/sql/catalyst/expressions/ParamLiteral.scala

```diff
@@ -61,10 +64,12 @@ class ParamLiteral(_value: Any, _dataType: DataType, val pos: Int)
     31 * (31 * Objects.hashCode(dataType)) + Objects.hashCode(pos)
   }
 
+  var considerUnequal = false
+
   override def equals(obj: Any): Boolean = obj match {
     case a: AnyRef if this eq a => true
     case pl: ParamLiteral =>
-      pl.dataType == dataType && pl.pos == pos
+      pl.dataType == dataType && pl.pos == pos && !(pl.considerUnequal || considerUnequal)
```

Comment:
I think this condition means that if either of the ParamLiterals are marked as "unequal" i.e. participant in BCHJ params, then force them to be unequal. This will ensure same plan is not used for two different tokens in BCHJ (which probably also means that each of those will be cached separately like without tokenization for which I have asked clarification from @kneeraj)